ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,447 @@ let order_f (_, vm_rec) =
+  | _, (vm, vm_size, group) :: remaining_vms ->
+      debug "No breach plan: try to plan for VM (%s %s %s)." (Ref.string_of vm)
+        (Db.VM.get_name_label ~__context ~self:vm)
+        (Db.VM_group.get_name_label ~__context ~self:group) ;
+      let no_resident_hosts =
+        VMGroupMap.find group pool_state.grp_no_resident_hosts_queue
+      in
+
+      let resident_hosts_cnt =
+        VMGroupMap.find group pool_state.grp_resident_hosts_count_map
+      in
+      info "No breach plan: total_hosts: %d, resident_hosts_cnt: %d."
+        total_hosts resident_hosts_cnt ;
+
+      if resident_hosts_cnt < 2 then (
+        debug
+          "No breach plan: there are less than 2 hosts has running VM in the \
+           same anti-affinity group, and there are still host(s) which has 0 \
+           running VMs, try to plan for it." ;
+        match select_host_for_no_breach_plan vm vm_size no_resident_hosts with
+        | None ->
+            info
+              "No breach plan: failed to select host on any of the no resident \
+               hosts, skip it, continue with the next VM." ;
+            compute_no_breach_plan ~__context total_hosts pool_state
+              remaining_vms mapping
+        | Some h ->
+            debug
+              "No breach plan: choose the no resident host with the minimal \
+               free memory which can run the vm: (%s)."
+              (Db.Host.get_name_label ~__context ~self:h) ;
+            compute_no_breach_plan ~__context total_hosts
+              (pool_state |> update_no_breach_plan_pool_state vm_size group h)
+              remaining_vms ((vm, h) :: mapping)
+      ) else (
+        debug
+          "No breach plan: no need to plan for the VM as the number of hosts \
+           which has running VMs from the same group is no less than 2, \
+           continue to plan for the next one." ;
+        compute_no_breach_plan ~__context total_hosts pool_state remaining_vms
+          mapping
+      )
+
+let anti_affinity_vms_increasing ~__context vms =
+  vms
+  |> List.filter_map (fun (vm, vm_size) ->
+         match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+         | Some (`AntiAffinity group) ->
+             Some (vm, vm_size, group)
+         | _ ->
+             None
+     )
+  |> List.sort (fun (_, a, _) (_, b, _) -> compare a b)
+
+(* Return an evacuation plan respecting VM anti-affinity rules: it is done in 3 phases:
+   1. Try to get a "spread evenly" plan for anti-affinity VMs, and then a binpack plan for
+      the rest of VMs. Done if every VM got planned, otherwise continue.
+   2. Try to get a "no breach" plan for anti-affinity VMs, and then a binpack plan for the
+      rest of VMs. Done if every VM got planned, otherwise continue.
+   3. Carry out a binpack plan ignoring VM anti-affinity. *)
+let compute_anti_affinity_evacuation_plan ~__context total_hosts hosts vms =
   let config =
     {Binpack.hosts; vms; placement= []; total_hosts; num_failures= 1}
```

Comment:
Why is `num_failures` 1 , or is it just from 1?

Reply:
I think it is used by binpack for HA plan only, not used for "host evacuation".